From c776fb13f01f088bec13c096e4b0a8cdb720407a Mon Sep 17 00:00:00 2001
From: linmin <linmin@eswincomputing.com>
Date: Wed, 27 Nov 2024 15:01:03 +0800
Subject: [PATCH 295/652] WIN2030-14758:fix:split dmabuf

Changelogs:
1.Determine the exp_name of dmabuf, if it's linux,cma then use different struct type
  to get it's sgt.

Signed-off-by: linmin <linmin@eswincomputing.com>
Change-Id: I120e2e900ae57b2fd145c31f513c32954c6003ac
---
 .../dmabuf-heap-import-helper.c               | 20 +++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
index 2a2877f848b25..d9ba68d918ef0 100644
--- a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
+++ b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
@@ -47,6 +47,13 @@ struct dma_heap_attachment {
 	bool mapped;
 };
 
+struct dma_heap_attachment_cma {
+	struct device *dev;
+	struct sg_table table;
+	struct list_head list;
+	bool mapped;
+};
+
 static int dmabuf_heap_add_buf_handle(struct dmaheap_file_private *prime_fpriv,
 				    struct dma_buf *dma_buf, uint64_t handle)
 {
@@ -893,6 +900,7 @@ static int eswin_get_split_dmabuf(struct eswin_split_buffer *split_buffer)
 	 struct sg_table *orig_splitted_sgt = &split_buffer->orig_sg_table;
 	struct dma_buf_attachment *attach;
 	struct dma_heap_attachment *a;
+	struct dma_heap_attachment_cma *cma_a;
 	struct sg_table *par_origin_table;
 	int out_nents[1];
 	struct scatterlist *sg;
@@ -914,8 +922,16 @@ static int eswin_get_split_dmabuf(struct eswin_split_buffer *split_buffer)
 		return ret;
 	}
 
-	a = (struct dma_heap_attachment *)attach->priv;
-	par_origin_table = a->table;
+	if (unlikely(!strcmp("linux,cma", par_dma_buf->exp_name)))
+	{
+		cma_a = (struct dma_heap_attachment_cma *)attach->priv;
+		par_origin_table = &cma_a->table;
+	}
+	else {
+		a = (struct dma_heap_attachment *)attach->priv;
+		par_origin_table = a->table;
+	}
+
 	#ifdef PRINT_ORIGINAL_SPLITTERS
 	{
 		pr_debug("%s:parent[0x%px]:sgt->orig_nents=%d, nents=%d, sg_nents(in) %d\n",
-- 
2.51.0

